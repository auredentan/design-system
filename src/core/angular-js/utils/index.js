import './lx-depth_service';
import './lx-enter-keypress';
import './lx-event-scheduler_service';
import './lx-utils_service';
