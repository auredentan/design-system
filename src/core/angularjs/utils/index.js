/* eslint-disable import/no-unassigned-import */

import './depth_service';
import './enter-keypress_directive';
import './event-scheduler_service';
import './focus-on-init_directive';
import './focus-trap_service';
import './utils_service';
