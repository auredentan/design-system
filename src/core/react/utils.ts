import React from 'react';

import get from 'lodash/get';
import has from 'lodash/has';
import isEmpty from 'lodash/isEmpty';
import isString from 'lodash/isString';

/////////////////////////////

type Component = string | React.FC<any> | React.PureComponent<any, any> | React.Component<any, any>;

/////////////////////////////

/**
 * Check if a ReactElement is of the given type.
 *
 * @param  {React.ReactElement} el   The ReactElement we want to check the type of.
 * @param  {string|Component}   type The type we want to check if the ReactElement is of.
 * @return {boolean}            If the ReactElement is of the given type or not.
 */
function isElementOfType(el: React.ReactElement, type: Component): boolean {
    if (!isString(type) && has(type, 'displayName')) {
        type = get(type, 'displayName', get(type, '_reactInternalFiber.elementType.name', 'Component'));
    }

    if (isEmpty(type) && !isString(type)) {
        throw new Error(
            `The type you want to check is not valid. Waiting a component or a string, got ${type.toString()}`,
        );
    }

    if (type === 'text') {
        return isString(el) || el.type === 'span';
    }

    return get(el.type, 'name') === type;
}

/**
 * Check if a ReactElement is a text (i.e. either a pure text node or a <span>).
 * Simply check if the ReactElement is a string or if its type is 'span'.
 *
 * @param  {React.ReactElement} el The ReactElement to check if it's a text.
 * @return {boolean}            If the ReactElement is a text or not.
 */
function isElementText(el: React.ReactElement): boolean {
    return isElementOfType(el, 'text');
}

/////////////////////////////

export { Component, isElementOfType, isElementText };
