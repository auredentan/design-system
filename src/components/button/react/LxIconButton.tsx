import React, { Children, cloneElement, Fragment } from 'react';

import classNames from 'classnames';

import get from 'lodash/get';
import isEmpty from 'lodash/isEmpty';

import { LxIcon } from 'LumX';
import { isElementOfType, unwrapFragment } from 'LumX/core/react/utils';

import {
    CLASSNAME as LXBUTTON_CLASSNAME,
    Color,
    Colors,
    Emphasis,
    Emphasises,
    LxButton,
    LxButtonProps,
    Size,
    Sizes,
    Theme,
    Themes,
} from './LxButton';

/////////////////////////////

enum Variants {
    icon = 'icon',
}
type Variant = Variants;

/////////////////////////////

/**
 * Defines the props of the component.
 */
interface IProps extends LxButtonProps {
    /**
     * The <LxIconButton> should never have the `variant` prop as this prop is forced to 'icon' in the <LxButton>.
     */
    variant?: never;
}
type LxIconButtonProps = IProps;

/////////////////////////////
//                         //
//    Public attributes    //
//                         //
/////////////////////////////

/**
 * The default class name and classes prefix for this component.
 *
 * @type {string}
 * @constant
 * @readonly
 */
const CLASSNAME: string = `${LXBUTTON_CLASSNAME}__icon`;

/////////////////////////////
//                         //
//    Private functions    //
//                         //
/////////////////////////////

/**
 * Validate the <LxIconButton> component props and children.
 * Also, sanitize, cleanup and format the children and return the processed ones.
 *
 * @param  {LxIconButtonProps}  props The children and props of the <LxButton> component.
 * @return {React.ReactNode} The processed children of the component.
 */
function _validate({ children, variant }: LxIconButtonProps): React.ReactNode {
    const newChildren: React.ReactNode = unwrapFragment(children);

    const childrenCount: number = Children.count(newChildren);
    if (childrenCount === 0) {
        throw new Error('Your <LxIconButton> must have at least 1 child for the icon (got 0)!');
    }

    if (childrenCount > 1) {
        throw new Error(`You cannot have more than 1 child in a <LxIconButton> (got ${childrenCount})!`);
    }

    Children.forEach(
        newChildren,
        (child: any): void => {
            if (isElementOfType(child, LxIcon)) {
                return;
            }

            throw new Error(
                `You can only have a <LxIcon> child in a <LxIconButton> (got '${get(child.type, 'name', child.type) ||
                    `'${child}'`}')!`,
            );
        },
    );

    if (!isEmpty(variant)) {
        console.warn(
            `You shouldn't pass the 'variant' prop in a <LxIconButton> as it's forced to 'icon' (got '${variant}')!`,
        );
    }

    return newChildren;
}

/////////////////////////////

/**
 * Displays an icon button.
 * It's like a <LxButton> but displays an icon instead of a label in the body of the button.
 *
 * Note that you cannot use the `variant` prop in this component.
 *
 * @see {@link LxButton} for more information on <LxButton>.
 *
 * @return {JSX.Element} The <LxIconButton> component.
 */
const LxIconButton: React.FC<LxIconButtonProps> = ({ children, ...props }: LxIconButtonProps): JSX.Element => {
    children = _validate({ children, ...props });

    return (
        <LxButton {...props} variant={Variants.icon}>
            {/* [XXX] Clement: Type of `child` should be React.ReactElement<LxIconProps>, but I didn't managed to make it work. */}
            {Children.map(children, (child: any) => {
                if (get(child.type, 'name') !== LxIcon.displayName) {
                    return <Fragment />;
                }

                return cloneElement(child, {
                    className: classNames(get(child.props, 'className', ''), CLASSNAME),
                });
            })}
        </LxButton>
    );
};
LxIconButton.displayName = 'LxIconButton';

/////////////////////////////

export {
    CLASSNAME,
    Color,
    Colors,
    Emphasis,
    Emphasises,
    LxIconButton,
    LxIconButtonProps,
    Size,
    Sizes,
    Theme,
    Themes,
    Variant,
    Variants,
};
