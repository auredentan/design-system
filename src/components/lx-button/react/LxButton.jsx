import React from 'react';
import classNames from 'classnames';

import { handleBasicClasses } from '../../../core/utils';
import '../style/lx-button.scss';

const LxButtonRoot = ({ href, children, ...props }) => {
    return href ? (
        <a href={href} {...props}>
            {children}
        </a>
    ) : (
        <button {...props}>{children}</button>
    );
};

const LxButton = ({
    children,
    className,
    color = 'primary',
    variant = 'primary',
    size = 'm',
    theme = 'light',
    ...props
}) => (
    <LxButtonRoot
        className={classNames(className, handleBasicClasses({ color, size, theme, variant, prefix: 'lx-button' }))}
        {...props}
    >
        {children}
    </LxButtonRoot>
);

export { LxButton };
