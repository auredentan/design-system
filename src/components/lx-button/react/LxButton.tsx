import React from 'react';
import classNames from 'classnames';

import '../style/lx-button.scss';

import { handleBasicClasses } from '../../../core/utils';
import { Color, Theme, Size, Variant } from 'components';

interface LxButtonRootProps {
    /** Basic react `className` prop. */
    className?: string;
    /** The `href` to reach if there is one. */
    href?: string;
    /** The `target` html property. */
    target?: string;
}

/**
 * Conditionnaly adds `a` or `button` html tag whether there is an `href` attribute or not.
 */
const LxButtonRoot: React.FC<LxButtonRootProps> = ({ href, children, target, className, ...props }) => {
    return href ? (
        <a href={href} target={target} className={className} {...props}>
            {children}
        </a>
    ) : (
        <button className={className} {...props}>
            {children}
        </button>
    );
};

export interface LxButtonProps extends LxButtonRootProps {
    /** The button color which must be defined by `lx-button--${color}` css class. */
    color?: Color;
    /** The button size which must be defined by `lx-button--${size}` css class. */
    size?: Size;
    /** The button theme which must be defined by `lx-button--${theme}` css class. */
    theme?: Theme;
    /** The button variant which must be defined by `lx-type--${variant}` css class. */
    variant?: Variant;
}

/**
 * Displays a button. If `href` property is set, it will display an `a` tag, if not
 * it will use `button` tag instead.
 */
export const LxButton: React.FC<LxButtonProps> = ({
    children,
    className,
    color = 'primary',
    variant = 'primary',
    size = 'm',
    theme = 'light',
    ...props
}) => (
    <LxButtonRoot
        className={classNames(className, handleBasicClasses({ color, size, theme, variant, prefix: 'lx-button' }))}
        {...props}
    >
        {children}
    </LxButtonRoot>
);
