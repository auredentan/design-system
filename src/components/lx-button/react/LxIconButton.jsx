import React from 'react';
import classNames from 'classnames';

import { LxButton } from './LxButton';
import '../style/lx-button.scss';

export const LxIconButton = ({ children, ...props }) => (
    <LxButton className="lx-button--shape-circled" {...props}>
        {React.Children.map(children, (child) =>
            React.cloneElement({
                ...child,
                props: {
                    ...child.props,
                    className: classNames(child.props.className, 'lx-button__icon'),
                },
            }),
        )}
    </LxButton>
);
