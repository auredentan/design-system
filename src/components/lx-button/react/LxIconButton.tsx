import React, { Children, cloneElement } from 'react';

import classNames from 'classnames';

import '../style/lx-button.scss';

import { ILxButtonProps, LxButton } from './LxButton';

export const LxIconButton: React.FC<ILxButtonProps> = ({ children, ...props }) => (
    <LxButton className="lx-button--shape-circled" {...props}>
        {Children.map(children, (child: any) =>
            cloneElement(child, {
                className: classNames(child.props.className, 'lx-button__icon'),
            }),
        )}
    </LxButton>
);
