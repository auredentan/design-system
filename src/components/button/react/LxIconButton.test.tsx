import React, { Fragment } from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { build, fake, oneOf } from 'test-data-bot';

import { LxIcon } from 'LumX';
import { ICommonSetup } from 'LumX/core/testing/utils.test';
import { mdiPlus } from 'LumX/icons';

import { LxButton, Variants } from './LxButton';
import { CLASSNAME, Emphasises, LxIconButton, LxIconButtonProps, Sizes, Themes } from './LxIconButton';

/////////////////////////////

/**
 * Define the overriding properties waited by the `setup` function.
 */
type ISetupProps = Partial<LxIconButtonProps>;

/**
 * Defines what is returned by the setup function.
 */
interface ISetup extends ICommonSetup {
    /**
     * The properties of the tested component.
     */
    props: ISetupProps;

    /**
     * The <LxButton> element that is used as a wrapper for the children of the <LxIconButton>.
     */
    button: ShallowWrapper;

    /**
     * The <LxIcon> icon(s) of the <LxIconButton>.
     */
    icon: ShallowWrapper;
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param  {ISetupProps} props The props to use to override the default props of the component.
 * @return {ISetup}            An object with the props, the component wrapper and some shortcut to some element inside of
 *                             the component.
 */
const setup = ({ ...propsOverrides }: ISetupProps = {}): ISetup => {
    const props: LxIconButtonProps = {
        children: <LxIcon icon={mdiPlus} />,
        ...propsOverrides,
    };

    const wrapper: ShallowWrapper = shallow(<LxIconButton {...props} />);

    return {
        button: wrapper.find('LxButton'),

        icon: wrapper.find('LxIcon'),

        props,
        wrapper,
    };
};

describe(`<${LxIconButton.displayName}>`, () => {
    // 1. Test render via snapshot (default state of component).
    describe('Snapshots and structure', (): void => {
        it('should render correctly an icon button', (): void => {
            const { button, icon, wrapper }: ISetup = setup();
            expect(wrapper).toMatchSnapshot();

            expect(button.exists()).toEqual(true);
            expect(icon.exists()).toEqual(true);
            expect(icon.length).toEqual(1);
            expect(icon.prop('className')).toContain(CLASSNAME);
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.
    describe('Props', (): void => {
        it('should use default props', (): void => {
            const { button }: ISetup = setup();

            expect(button.prop('variant')).toEqual(Variants.icon);
        });

        it("should user 'icon' `variant` whatever the given prop is", (): void => {
            // Disable the display of the warn message in the console.
            global.console.warn = jest.fn();

            const testedProp: string = 'variant';
            const modifiedProps: ISetupProps = {
                [testedProp]: Variants.icon,
            };

            let { button }: ISetup = setup(modifiedProps);

            expect(button.prop(testedProp)).toEqual(Variants.icon);

            /////////////////////////////

            modifiedProps[testedProp] = Variants.button;

            ({ button } = setup(modifiedProps));

            expect(button.prop(testedProp)).toEqual(Variants.icon);

            // @ts-ignore
            global.console.warn.mockRestore();
        });

        it(`should forward any <${LxButton.displayName}> prop (except \`variant\`)`, (): void => {
            const modifiedPropsBuilder: () => ISetupProps = build('props').fields({
                color: fake((fakeData) => fakeData.commerce.color()),
                emphasis: oneOf(...Object.values(Emphasises)),
                size: oneOf(...Object.values(Sizes)),
                theme: oneOf(...Object.values(Themes)),
            });

            const modifiedProps: ISetupProps = modifiedPropsBuilder();
            if (modifiedProps.emphasis !== Emphasises.high) {
                delete modifiedProps.theme;
            }
            const { button }: ISetup = setup(modifiedProps);

            Object.keys(modifiedProps).forEach((prop: string) => {
                expect(button.prop(prop)).toEqual(modifiedProps[prop]);
            });
        });

        it('should forward any CSS class', (): void => {
            const testedProp: string = 'className';
            const modifiedProps: ISetupProps = {
                [testedProp]: 'component component--is-tested',
            };

            const { button }: ISetup = setup(modifiedProps);

            expect(button.prop(testedProp)).toContain(modifiedProps[testedProp]);
        });

        it('should forward any other props', (): void => {
            const testedProp: string = 'winter';
            const modifiedProps: ISetupProps = {
                [testedProp]: 'is coming',
            };

            const { button }: ISetup = setup(modifiedProps);

            expect(button.prop(testedProp)).toEqual(modifiedProps[testedProp]);
        });
    });

    /////////////////////////////

    // 3. Test events.
    describe('Events', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', (): void => {
        beforeEach(
            (): void => {
                try {
                    // If `console.warn` has been mocked at least one, this exists. So disable TS here.
                    // @ts-ignore
                    global.console.warn.mockRestore();
                } catch (exception) {
                    // Nothing to do here.
                }
            },
        );

        it('should fail when no child is given', (): void => {
            expect(
                (): void => {
                    setup({ children: null });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it('should fail when more than 1 children are given', (): void => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            expect(
                (): void => {
                    setup({ children });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it(`should fail when anything else than <${LxIcon.displayName}> is given as child`, (): void => {
            const children: React.ReactNode = <div>toto</div>;

            expect(
                (): void => {
                    setup({ children });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it(`should warn the user when rendering a <${LxIconButton.displayName}> with a \`variant\``, (): void => {
            global.console.warn = jest.fn();

            // We know that a <LxIconButton> cannot receive a `variant`, but for the purpose of the test we ignore this.
            // @ts-ignore
            setup({ variant: Variants.icon });
            expect(global.console.warn).toHaveBeenCalled();

            // @ts-ignore
            global.console.warn.mockClear();

            // We know that a <LxIconButton> cannot receive a `variant`, but for the purpose of the test we ignore this.
            // @ts-ignore
            setup({ variant: Variants.button });
            expect(global.console.warn).toHaveBeenCalled();
        });
    });

    /////////////////////////////

    // 5. Test state.
    describe('State', (): void => {
        // Nothing to do here.
    });
});
