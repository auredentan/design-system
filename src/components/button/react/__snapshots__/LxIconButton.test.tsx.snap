// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LxIconButton> Conditions should fail when anything else than <LxIcon> is given as child 1`] = `"You can only have <LxIcon> children in <LxIconButton> (got <div>)!"`;

exports[`<LxIconButton> Conditions should fail when more than 1 children are given 1`] = `"<LxIconButton> must have exactly 1 child (got 2)!"`;

exports[`<LxIconButton> Conditions should fail when no child is given 1`] = `"<LxIconButton> must have exactly 1 child (got 0)!"`;

exports[`<LxIconButton> Snapshots and structure should render correctly an icon button 1`] = `
<LxButton
  variant="icon"
>
  <LxIcon
    className="lumx-button__icon"
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
</LxButton>
`;
